Add documentation for installing custom certificates to Dapr sidecar (#2572)
dapr/docs: daprdocs/content/en/operations/configuration/use-custom-certificates.md

---
type: docs
title: "How-To: Use custom certificates in the Dapr sidecar"
linkTitle: "Use custom certificates"
weight: 6500
description: "Configure the Dapr sidecar container to trust custom certificates"
---

> The Dapr sidecar can be configured to trust custom certificates for communicating with external services. This is useful in scenarios where a self-signed certificate needs to be trusted, e.g., using the HTTP binding, configuring an outbound proxy for the sidecar, etc.

Reviewer: Can you add a sentence here explaining that you can add both leaf certificates as well as custom root CA certificates?

Author: Sure!

Author: Please let me know if it looks ok, or should I reword something.

{{< tabs Self-hosted Kubernetes >}}

{{% codetab %}}

> When the sidecar is running as a process, certificates must be installed on the host operating system. 

Reviewer: “When the sidecar is not running inside a container”

> When the sidecar is running as a container,

Reviewer: end with :

1. Certificates must be available to the sidecar container. This can be configured using volume mounts.
1. The environment variable `SSL_CERT_DIR` must be set in the sidecar container, pointing to the directory containing the certificates.

> 1. Additionally for Windows containers, the container needs to be run with administrator privileges. This is required by the Windows entrypoint script to install the certificates.

Reviewer: “1. For Windows containers only, the container needs to run with administrator privileges to be able to install the certificates.”

> This is an example using Docker compose:

Reviewer: “Docker Compose” -> Uppercase C

```yaml
version: '3'
services:
  dapr-sidecar:
    image: "daprio/daprd:edge"
    command: [
      "./daprd",
     "-app-id", "myapp",
     "-app-port", "3000",
     ]
    volumes:
        - "./components/:/components"
        - "./certificates:/certificates" # (STEP 1) Mount the certificates folder to the sidecar container
    environment:
      - "SSL_CERT_DIR=/certificates" # (STEP 2) Set the environment variable to the path of the certificates folder

>     user: ContainerAdministrator # (STEP 3) Optional, required for Windows containers

Reviewer: I would change this to be commented by default. Then add before a comment that says “# Uncomment this for Windows containers”

```

{{% /codetab %}}

{{% codetab %}}

On Kubernetes, 
1. Certificates must be available to the sidecar container using a volume mount.
1. The environment variable `SSL_CERT_DIR` must be set in the sidecar container, pointing to the directory containing the certificates.

This is an example deployment YAML:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: myapp
  namespace: default
  labels:
    app: myapp
spec:
  replicas: 1
  selector:
    matchLabels:
      app: myapp
  template:
    metadata:
      labels:
        app: myapp
      annotations:
        dapr.io/enabled: "true"
        dapr.io/app-id: "myapp"
        dapr.io/app-port: "8000"
        dapr.io/volume-mounts-rw: "certificates-vol:/tmp/certificates" # (STEP 1) Mount the certificates folder to the sidecar container
        dapr.io/env: "SSL_CERT_DIR=/tmp/certificates" # (STEP 2) Set the environment variable to the path of the certificates folder
    spec:
      volumes:
        - name: certificates-vol
          hostPath:
            path: /certificates
...
```

> Note, when using Windows container, the sidecar injector injects the sidecar with admin privileges. This is required by the Windows entrypoint script to install the certificates.

Reviewer: I don’t think it’s necessary to explain about entry points. I’d just say:

```
**Note for Windows containers:** When using Windows containers, the sidecar container is started with admin privileges, which is required to install the certificates. This does not apply to Linux containers.
```

---
type: docs
title: "How-To: Use custom certificates in the Dapr sidecar"
linkTitle: "Use custom certificates"
weight: 6500
description: "Configure the Dapr sidecar container to trust custom certificates"
---

> The Dapr sidecar can be configured to trust custom certificates for communicating with external services. This is useful in scenarios where a self-signed certificate needs to be trusted, e.g., using the HTTP binding, configuring an outbound proxy for the sidecar, etc. Note, both custom CA certificates and leaf certificates are supported.

Reviewer 2: ```suggestion
The Dapr sidecar can be configured to trust custom certificates for communicating with external services. This is useful in scenarios where a self-signed certificate needs to be trusted, e.g., using an HTTP binding, configuring an outbound proxy for the sidecar, etc. Both custom CA certificates and leaf certificates are supported.
```

{{< tabs Self-hosted Kubernetes >}}

{{% codetab %}}

When the sidecar is not running inside a container, certificates must be directly installed on the host operating system. 

When the sidecar is running as a container:
1. Certificates must be available to the sidecar container. This can be configured using volume mounts.
1. The environment variable `SSL_CERT_DIR` must be set in the sidecar container, pointing to the directory containing the certificates.
1. For Windows containers only, the container needs to run with administrator privileges to be able to install the certificates.

> This is an example using Docker Compose:

Reviewer 2: ```suggestion
Below is an example that uses Docker Compose:
```

```yaml
version: '3'
services:
  dapr-sidecar:
    image: "daprio/daprd:edge"
    command: [
      "./daprd",
     "-app-id", "myapp",
     "-app-port", "3000",
     ]
    volumes:
        - "./components/:/components"
        - "./certificates:/certificates" # (STEP 1) Mount the certificates folder to the sidecar container
    environment:
      - "SSL_CERT_DIR=/certificates" # (STEP 2) Set the environment variable to the path of the certificates folder
    # Uncomment this for Windows containers
    # user: ContainerAdministrator
```

{{% /codetab %}}

{{% codetab %}}

On Kubernetes:
1. Certificates must be available to the sidecar container using a volume mount.
1. The environment variable `SSL_CERT_DIR` must be set in the sidecar container, pointing to the directory containing the certificates.

> This is an example deployment YAML:

Reviewer 2: ```suggestion
Below is an example of a deployment YAML:
```

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: myapp
  namespace: default
  labels:
    app: myapp
spec:
  replicas: 1
  selector:
    matchLabels:
      app: myapp
  template:
    metadata:
      labels:
        app: myapp
      annotations:
        dapr.io/enabled: "true"
        dapr.io/app-id: "myapp"
        dapr.io/app-port: "8000"

>         dapr.io/volume-mounts-rw: "certificates-vol:/tmp/certificates" # (STEP 1) Mount the certificates folder to the sidecar container

Reviewer 3: ```suggestion
        dapr.io/volume-mounts: "certificates-vol:/tmp/certificates" # (STEP 1) Mount the certificates folder to the sidecar container
```

Should this only be in read only mode as a best practice?

Author: Agreed, changing it.

        dapr.io/env: "SSL_CERT_DIR=/tmp/certificates" # (STEP 2) Set the environment variable to the path of the certificates folder
    spec:
      volumes:
        - name: certificates-vol
          hostPath:
            path: /certificates
...
```

> **Note for Windows containers:** When using Windows containers, the sidecar container is started with admin privileges, which is required to install the certificates. This does not apply to Linux containers.

Reviewer 3: add it as a warning alert ??

Author: Should we make this info instead? It is less of a warning (no action item for the user), and more of an FYI.

---
type: docs

> title: "How-To: Use custom certificates in the Dapr sidecar"

Reviewer 4: Why do you use "Custom" in the title? What is custom about this? Should this just be how to load certificates into the Dapr sidecar?

Author: Well, the original feature was to be able to install "extra"/"custom" certificates in the sidecar (similar to bring-your-own-certs), and hence the name. But yes, we can drop the "custom". Editing it.

linkTitle: "Use custom certificates"
weight: 6500
description: "Configure the Dapr sidecar container to trust custom certificates"
---

> The Dapr sidecar can be configured to trust custom certificates for communicating with external services. This is useful in scenarios where a self-signed certificate needs to be trusted, e.g., using an HTTP binding, configuring an outbound proxy for the sidecar, etc. Both custom CA certificates and leaf certificates are supported.

Reviewer 4: ```suggestion
The Dapr sidecar can be configured to trust custom certificates for communicating with external services. This is useful in scenarios where a self-signed certificate needs to be trusted. For example, using an HTTP binding or configuring an outbound proxy for the sidecar. Both custom certificate authority (CA) certificates and leaf certificates are supported.
```

{{< tabs Self-hosted Kubernetes >}}

{{% codetab %}}

When the sidecar is not running inside a container, certificates must be directly installed on the host operating system. 

When the sidecar is running as a container:
1. Certificates must be available to the sidecar container. This can be configured using volume mounts.

> 1. The environment variable `SSL_CERT_DIR` must be set in the sidecar container, pointing to the directory containing the certificates.

Reviewer 4: Can you update https://v1-8.docs.dapr.io/reference/environment/ with this environment variable with a description and who sets this value

> 1. The environment variable `SSL_CERT_DIR` must be set in the sidecar container, pointing to the directory containing the certificates.

Reviewer 4: Does this load ALL certificates in this directory? And what specifically is this looking for? .cer files or some other format of certs? Do the certificates have to have a particular file type or extension to be loaded? Can you specify details to help the user understand the certs deployment files?

And in Windows this directory approach still applies, even though Windows uses a cert store?

Author: For the first part, I have added the information at the end of the document.

Yes, it is also supported on Windows. We have a custom logic in place to read from this directory on startup and add cert to the root cert store. Check this out - https://github.com/dapr/dapr/pull/4741

> 1. For Windows containers only, the container needs to run with administrator privileges to be able to install the certificates.

Reviewer 3: ```suggestion
1. For Windows containers, the container needs to run with administrator privileges to be able to install the certificates.
```

> Below is an example that uses Docker Compose:

Reviewer 4: Explain this some more. What is this example about to show me, since I am expected to use this locally on my dev machine, yes?

```suggestion
Below is an example that uses Docker Compose to XXXXX.
```

```yaml
version: '3'
services:
  dapr-sidecar:

>     image: "daprio/daprd:edge"

Reviewer 3: add comment for making users aware that they need to use versions 1.8 or above ....

Author: So this doc is a part of v1.8, does not that make it implicit? Anything that's present in doc version X and not X-1, will need a min version X.

Author: Anyway I can add a tiny comment to be safe.

>     image: "daprio/daprd:edge" # dapr version must be at least 1.8

Reviewer 4: ```suggestion
    image: "daprio/daprd:edge" # dapr version must be at least v1.8
```

    command: [
      "./daprd",
     "-app-id", "myapp",
     "-app-port", "3000",
     ]
    volumes:
        - "./components/:/components"
        - "./certificates:/certificates" # (STEP 1) Mount the certificates folder to the sidecar container
    environment:
      - "SSL_CERT_DIR=/certificates" # (STEP 2) Set the environment variable to the path of the certificates folder

>     # Uncomment this for Windows containers

Reviewer 4: ```suggestion
    # Uncomment the line below for Windows containers
```

    # user: ContainerAdministrator
```

{{% /codetab %}}

{{% codetab %}}

On Kubernetes:
1. Certificates must be available to the sidecar container using a volume mount.
1. The environment variable `SSL_CERT_DIR` must be set in the sidecar container, pointing to the directory containing the certificates.

> Below is an example of a deployment YAML:

Reviewer 4: ```suggestion
The YAML below is an example of a deployment:
```

Reviewer 4: Say a little more on what this does.

...
```

> > ℹ️ **Note for Windows containers:** When using Windows containers, the sidecar container is started with admin privileges, which is required to install the certificates. This does not apply to Linux containers.

Reviewer 3: ```suggestion
{{% alert title="Note" color="primary" %}}
When using Windows containers, the sidecar container is started with admin privileges, which is required to install the certificates. This does not apply to Linux containers.
{{% /alert %}}
```

See example note in https://docs.dapr.io/operations/components/component-scopes/

Author: Nice, switching.
